File: CS/IssueService/IssuesService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace InfiniteAsyncSourceAdvancedSample {
    public static class IssuesService {
        #region helpers
        static object SyncObject = new object();
        static Lazy<IssueData[]> AllIssues = new Lazy<IssueData[]>(() => {
            var date = DateTime.Today;
            var rnd = new Random(0);
            return Enumerable.Range(0, 100000)
                .Select(i => {
                    date = date.AddSeconds(-rnd.Next(20 * 60));
                    return new IssueData(
                        subject: OutlookDataGenerator.GetSubject(),
                        user: OutlookDataGenerator.GetFrom(),
                        created: date,
                        votes: rnd.Next(100),
                        tags: OutlookDataGenerator.GetTags());
                }).ToArray();
        });
        #endregion

        public async static Task<IssueData[]> GetIssuesAsync(int page, int pageSize, IssueSortOrder sortOrder, IssueFilter filter) {
            await Task.Delay(300);
            CheckRestrictions(sortOrder, filter);
            var issues = SortIssues(sortOrder, AllIssues.Value);
            if(filter != null)
                issues = FilterIssues(filter, issues);
            return issues.Skip(page * pageSize).Take(pageSize).ToArray();
        }

        static void CheckRestrictions(IssueSortOrder sortOrder, IssueFilter filter) {
            if(filter == null)
                return;
            if((sortOrder == IssueSortOrder.Hot || sortOrder == IssueSortOrder.Week)
                && (filter.CreatedFrom != null || filter.CreatedTo != null || filter.MaxVotes != null || filter.MinVotes != null))
                throw new InvalidOperationException("Restrictions violation");

            if((filter.MaxVotes != null || filter.MinVotes != null)
                && !(sortOrder == IssueSortOrder.VotesAscending || sortOrder == IssueSortOrder.VotesDescending))
                throw new InvalidOperationException("Restrictions violation");
        }

        public async static Task<string[]> GetTagsAsync() {
            await Task.Delay(300);
            return OutlookDataGenerator.Tags;
        }


        #region filter
        static IEnumerable<IssueData> FilterIssues(IssueFilter filter, IEnumerable<IssueData> issues) {
            if(filter.CreatedFrom != null) {
                issues = issues.Where(x => x.Created >= filter.CreatedFrom.Value);
            }
            if(filter.CreatedTo != null) {
                issues = issues.Where(x => x.Created < filter.CreatedTo.Value);
            }
            if(filter.MinVotes != null) {
                issues = issues.Where(x => x.Votes >= filter.MinVotes.Value);
            }
            if(filter.MaxVotes != null) {
                issues = issues.Where(x => x.Votes <= filter.MaxVotes.Value);
            }
            if(!string.IsNullOrEmpty(filter.Tag)) {
                issues = issues.Where(x => x.Tags.Contains(filter.Tag));
            }
            return issues;
        }
        #endregion

        #region sort
        static IEnumerable<IssueData> SortIssues(IssueSortOrder sortOrder, IEnumerable<IssueData> issues) {
            switch(sortOrder) {
            case IssueSortOrder.Default:
                return issues
                    .Where(x => x.Created > DateTime.Today.AddDays(-2) && x.Votes > 70)
                    .OrderByDescending(x => x.Created);
            case IssueSortOrder.Hot:
                return issues
                    .Where(x => x.Created > DateTime.Today.AddDays(-2) && x.Votes > 70)
                    .OrderByDescending(x => x.Votes)
                    .ThenByDescending(x => x.Created);
            case IssueSortOrder.Week:
                return issues
                    .Where(x => x.Created > DateTime.Today.AddDays(-7) && x.Votes > 90)
                    .OrderByDescending(x => x.Votes)
                    .ThenByDescending(x => x.Created);

            case IssueSortOrder.CreatedAscending:
                return issues.OrderBy(x => x.Created);
            case IssueSortOrder.CreatedDescending:
                return issues.OrderByDescending(x => x.Created);

            case IssueSortOrder.VotesAscending:
                return issues
                    .OrderBy(x => x.Votes)
                    .ThenByDescending(x => x.Created);
            case IssueSortOrder.VotesDescending:
                return issues
                    .OrderByDescending(x => x.Votes)
                    .ThenByDescending(x => x.Created);
            default:
                throw new InvalidOperationException();
            }
        } 
        #endregion
    }
}

File: VB/IssueService/IssuesService.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading.Tasks

Namespace InfiniteAsyncSourceAdvancedSample

    Public Module IssuesService

#Region "helpers"
        Private SyncObject As Object = New Object()

        Private AllIssues As Lazy(Of IssueData()) = New Lazy(Of IssueData())(Function()
            Dim [date] = Date.Today
            Dim rnd = New Random(0)
            Return Enumerable.Range(0, 100000).[Select](Function(i)
                [date] = [date].AddSeconds(-rnd.Next(20 * 60))
                Return New IssueData(subject:=GetSubject(), user:=GetFrom(), created:=[date], votes:=rnd.Next(100), tags:=GetTags())
            End Function).ToArray()
        End Function)

#End Region
        Public Async Function GetIssuesAsync(ByVal skip As Integer, ByVal take As Integer, ByVal sortOrder As IssueSortOrder, ByVal filter As IssueFilter) As Task(Of IssueData())
            Await Task.Delay(300).ConfigureAwait(False)
            CheckRestrictions(sortOrder, filter)
            Dim issues = SortIssues(sortOrder, AllIssues.Value)
            If filter IsNot Nothing Then issues = FilterIssues(filter, issues)
            Return issues.Skip(skip).Take(take).ToArray()
        End Function

        Private Sub CheckRestrictions(ByVal sortOrder As IssueSortOrder, ByVal filter As IssueFilter)
            If filter Is Nothing Then Return
            If(sortOrder = IssueSortOrder.Hot OrElse sortOrder = IssueSortOrder.Week) AndAlso (filter.CreatedFrom IsNot Nothing OrElse filter.CreatedTo IsNot Nothing OrElse filter.MaxVotes IsNot Nothing OrElse filter.MinVotes IsNot Nothing) Then Throw New InvalidOperationException("Restrictions violation")
            If(filter.MaxVotes IsNot Nothing OrElse filter.MinVotes IsNot Nothing) AndAlso Not(sortOrder = IssueSortOrder.VotesAscending OrElse sortOrder = IssueSortOrder.VotesDescending) Then Throw New InvalidOperationException("Restrictions violation")
        End Sub

        Public Async Function GetTagsAsync() As Task(Of String())
            Await Task.Delay(300)
            Return Tags
        End Function

#Region "filter"
        Private Function FilterIssues(ByVal filter As IssueFilter, ByVal issues As IEnumerable(Of IssueData)) As IEnumerable(Of IssueData)
            If filter.CreatedFrom IsNot Nothing Then
                issues = issues.Where(Function(x) x.Created >= filter.CreatedFrom.Value)
            End If

            If filter.CreatedTo IsNot Nothing Then
                issues = issues.Where(Function(x) x.Created < filter.CreatedTo.Value)
            End If

            If filter.MinVotes IsNot Nothing Then
                issues = issues.Where(Function(x) x.Votes >= filter.MinVotes.Value)
            End If

            If filter.MaxVotes IsNot Nothing Then
                issues = issues.Where(Function(x) x.Votes <= filter.MaxVotes.Value)
            End If

            If Not String.IsNullOrEmpty(filter.Tag) Then
                issues = issues.Where(Function(x) x.Tags.Contains(filter.Tag))
            End If

            Return issues
        End Function

#End Region
#Region "sort"
        Private Function SortIssues(ByVal sortOrder As IssueSortOrder, ByVal issues As IEnumerable(Of IssueData)) As IEnumerable(Of IssueData)
            Select Case sortOrder
                Case IssueSortOrder.Default
                    Return issues.Where(Function(x) x.Created > Date.Today.AddDays(-2) AndAlso x.Votes > 70).OrderByDescending(Function(x) x.Created)
                Case IssueSortOrder.Hot
                    Return issues.Where(Function(x) x.Created > Date.Today.AddDays(-2) AndAlso x.Votes > 70).OrderByDescending(Function(x) x.Votes).ThenByDescending(Function(x) x.Created)
                Case IssueSortOrder.Week
                    Return issues.Where(Function(x) x.Created > Date.Today.AddDays(-7) AndAlso x.Votes > 90).OrderByDescending(Function(x) x.Votes).ThenByDescending(Function(x) x.Created)
                Case IssueSortOrder.CreatedAscending
                    Return issues.OrderBy(Function(x) x.Created)
                Case IssueSortOrder.CreatedDescending
                    Return issues.OrderByDescending(Function(x) x.Created)
                Case IssueSortOrder.VotesAscending
                    Return issues.OrderBy(Function(x) x.Votes).ThenByDescending(Function(x) x.Created)
                Case IssueSortOrder.VotesDescending
                    Return issues.OrderByDescending(Function(x) x.Votes).ThenByDescending(Function(x) x.Created)
                Case Else
                    Throw New InvalidOperationException()
            End Select
        End Function
#End Region
    End Module
End Namespace
